client/post-editor/post-editor.jsx

```diff
@@ -1013,17 +1014,268 @@ export const PostEditor = React.createClass( {
 		return editorMode;
 	},
 
+	isCursorInATag: function( content, cursorPosition ) {
+		const lastLtPos = content.lastIndexOf( '<', cursorPosition );
+		const lastGtPos = content.lastIndexOf( '>', cursorPosition );
+
+		if ( lastLtPos > lastGtPos ) {
+			// inside a tag that was opened, but not closed
+
+			// find what the tag is
+			const tagContent = content.substr( lastLtPos );
+			const tagMatch = tagContent.match( /<\s*(\/)?(\w+)/ );
+			if ( ! tagMatch ) {
+				return null;
+			}
+
+			const tagType = tagMatch[ 2 ];
+			const closingGt = tagContent.indexOf( '>' );
+			const isClosingTag = ! ! tagMatch[ 1 ];
+
+			return {
+				ltPos: lastLtPos,
+				gtPos: lastLtPos + closingGt + 1, // offset by one to get the position _after_ the character,
+				tagType,
+				isClosingTag,
+			};
+		}
+		return null;
+	},
+
+	addHTMLBookmarkInTextAreaContent: function() {
+		const textArea = this.editor._editor.getElement();
+
+		let htmlModeCursorStartPosition = textArea.selectionStart;
+		let htmlModeCursorEndPosition = textArea.selectionEnd;
+
+		// check if the cursor is in a tag and if so, adjust it
+		const isCursorStartInTag = this.isCursorInATag( textArea.value, htmlModeCursorStartPosition );
+		if ( isCursorStartInTag ) {
+			htmlModeCursorStartPosition = isCursorStartInTag.ltPos;
+		}
+
+		const isCursorEndInTag = this.isCursorInATag( textArea.value, htmlModeCursorEndPosition );
+		if ( isCursorEndInTag ) {
+			htmlModeCursorEndPosition = isCursorEndInTag.gtPos;
+		}
+
+		let mode = 'single';
+		if ( htmlModeCursorStartPosition !== htmlModeCursorEndPosition ) {
+			mode = 'range';
+		}
+
+		const bookMarkStart = '<span ' +
+			'data-mce-type="bookmark"	' +
+			'id="mce_SELREST_start" ' +
+			'data-mce-style="overflow:hidden;line-height:0px" ' +
+			'style="overflow:hidden;line-height:0px"' +
+			'>' +
+			'&#65279;' +
+			'</span>';
+
+		let bookMarkEnd = null;
```

Comment:
being that `bookMarkEnd` is null until modified in the conditional logic below nothing would show if we were to include it in the array and call `join` on the array.
If we were to do something similar with `textArea.value.slice( htmlModeCursorStartPosition, htmlModeCursorEndPosition )` then I think we could neaten up the construction of the `textArea.value`.

That might look like so:

```js
const bookMarkStart = '<span ... >';

let bookMarkEnd = null;
let bookMarkMiddle = null; // maybe a better name?

if ( mode === 'range' ) {
  bookMarkEnd = '<span ... >';
  bookMarkMiddle = textArea.value.slice( htmlModeCursorStartPosition, htmlModeCursorEndPosition );
}

textArea.value = [
  textArea.value.slice( 0, htmlModeCursorStartPosition ),
  bookMarkStart,
  bookMarkMiddle,
  bookMarkEnd,
  textArea.value.slice( htmlModeCursorEndPosition ),
].join( '' );
```

Would save the slight complexity of having 2 different array compositions to look at...

Reply:
It's good to have someone take a look on complicated logic like that. Makes more sense to use the above format as it is more readable and more condensed!